### PR TITLE
dataflow: fuel tank for sinks

### PR DESCRIPTION
**Pull request by ruchirK**
Touches #2806

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/3246)

### CODE REVIEW

**Comment by ruchirK** [comment]
I'd love a way to test the case where |diff| > 1 but I'm not sure how to coerce that with Materialize. any and all testing suggestions are welcome

**Comment by frankmcsherry** [comment]
I think something like `SELECT true FROM values () ()` should produce `(true, time, 2)`?

File: src/dataflow/src/sink/kafka.rs [hunk 1]
```diff
@@ -85,39 +89,65 @@ pub fn kafka<G>(
...
+                }
+            });
+
+            for _ in 0..KAFKA_SINK_FUEL {
+                let (encoded, count) = if let Some((encoded, count)) = encoded_buffer.take() {
+                    (encoded, count)
+                } else if queue.is_empty() {
+                    return;
+                } else {
+                    let (row, diff) = queue.pop_front().expect("queue known to be nonempty");
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:115]: Might be wrong, but can't this be `else if let Some((row, diff)) = queue.pop_front() { .. } else { return; }`?

File: src/dataflow/src/sink/kafka.rs [hunk 2]
```diff
@@ -85,39 +89,65 @@ pub fn kafka<G>(
...
-                        }
+                    (buf, diff.abs())
+                };
+
+                let record = BaseRecord::<&Vec<u8>, _>::to(&connector.topic).payload(&encoded);
+                if let Err((e, _)) = producer.inner().send(record) {
+                    error!("unable to produce in {}: {}", name, e);
+                    if let KafkaError::MessageProduction(RDKafkaError::QueueFull) = e {
+                        encoded_buffer = Some((encoded, count));
+                        return;
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:143]: I'm not familiar with the class of Kafka errors, but it looks like if we reach one other than `QueueFull` we will spam the heck out of everyone. Maybe best to return in that case too?

File: src/dataflow/src/sink/kafka.rs [hunk 3]
```diff
@@ -24,6 +26,8 @@ use expr::GlobalId;
 use interchange::avro::{DiffPair, Encoder};
 use repr::{RelationDesc, Row};
 
+static KAFKA_SINK_FUEL: usize = 10000;
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:29]: Can we make this either be part of the `KafkaSinkConnector` or an argument to `kafka`? I figure it is a reasonable knob to want to dial around, and having it be hardwired seems like something we'll want to fix at some point.

**Comment by benesch** [comment]
I'm mostly deferring to Frank on this review, but something I'm confused about is how timely knows to keep scheduling the sink operator when it has data buffered. Since the sink drains its input fully whenever it's scheduled, but might not send its entire queue to Kafka, is there a risk that timely will stop scheduling the sink if there are no new messages arriving, and we'll never flush the full queue?

**Comment by frankmcsherry** [comment]
Oh, right! There are definitely more issues here. I guess I spaced out:

1. The operator will want to re-activate itself after some amount of time, appropriate for letting the Kafka queue drain a bit. That could be right away, or it could be something else depending on what we believe about needing to chill out. Probably fine just to go again ASAP.
2. The operator will need to tell timely to keep it alive. Operators are closed when they have no inputs, hold no capabilities, and have not returned `true` from the last scheduling event (whose return value indicates whether they are incomplete). I generalized a bit of timely recently to make this easier for sinks to gain access to, and we'll need to do that here.

**Comment by ruchirK** [comment]
> I think something like `SELECT true FROM values () ()` should produce `(true, time, 2)`?

This worked! Added a test + verified that the code path was being hit by printing out diffs as we were writing them

File: src/dataflow/src/sink/kafka.rs (hunk 2 above)
Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs]: I'm not sure either which set of errors is worth retrying for vs which is not. this behavior preserves current functionality with the exception of QueueFull

**Comment by ruchirK** [comment]
I think I've addressed all the pr comments at this time. Tested by creating a large view w/ generate_series and sinking that and it works correctly

File: src/dataflow/src/sink/kafka.rs (hunk 2 above)
Comment (frankmcsherry) [reply src/dataflow/src/sink/kafka.rs]: I'm a bit confused about the error case. It seems like for `QueueFull` we return without decrementing the count (good), whereas for any other error we mark the record as sent and continue. There appear to be other sorts of errors like `OperationTimedOut` that we should retry, and others that we should probably do something a bit more assertive. I don't know that this PR is the right place to enumerate all of those, but we should probably make a note somewhere that this error means data loss occurred.

Comment (frankmcsherry) [reply src/dataflow/src/sink/kafka.rs]: Another option would just be to retry everything after 60s, or whatever, which has the property that transient errors can recover and persistent failures just turn in to memory leaks. I'm not sure which case worries us more, but my default guess would be avoid the data loss to start.

**Comment by ruchirK** [comment]
@frankmcsherry I added some more comments around the error handling approach / why I think its ok to only retry on QueueFull + tested with a large topic to validate that this does indeed make materialize more responsive. let me know if you think this needs anything else

**Comment by frankmcsherry** [comment]
Seems mergeable to me! I do have a moral preference for errors generating retries rather than dropping messages (in the vein of "at least once", though without that conviction). It seems like the Kafka folks don't agree, but I'm not sure the best forum to shake that out. Seems good!

**Pull request merged**